aips/aip-20.md

```diff
@@ -0,0 +1,467 @@
+1. Compute the lexicographical flag as `b[0] & 0x20 != 0`.
+1. Deserialize `[b[0] & 0x1f, b[1], ..., b[95]]` to `x` using `FormatFq2MscMsb`. If `x` is none, return none.
+1. Solve the curve equation with `x` for `y`. If no such `y` exists, return none.
+1. Let `y'` be `max(y,-y)` if the lexicographical flag is set, or `min(y,-y)` otherwise.
+1. Check if `(x,y')` is in the subgroup of order `r`. If not, return none.
+1. Return `(x,y')`.
+
+NOTE: other implementation(s) using this format: ark-bls12-381-0.4.0.
+
+#### `Gt` (implemented)
+The group $G_t$ in BLS12-381-based pairing $G_1 \times G_2 \rightarrow G_t$.
+It is a multiplicative subgroup of `Fq12`,
+with a prime order $r$ equal to 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001.
+(so `Fr` is the scalar field).
+The identity of `Gt` is 1.
+
+#### `FormatGt` (implemented)
+A serialization scheme for `Gt` elements.
+
+To serialize, it treats a `Gt` element `p` as an `Fq12` element and serialize it using `FormatFq12LscLsb`.
+
+To deserialize, it uses `FormatFq12LscLsb` to try deserializing to an `Fq12` element then test the membership in `Gt`.
+
+NOTE: other implementation(s) using this format: ark-bls12-381-0.4.0.
+
+#### `Fr` (implemented)
+The finite field $F_r$ that can be used as the scalar fields
+associated with the groups $G_1$, $G_2$, $G_t$ in BLS12-381-based pairing.
+
+#### `FormatFrLsb` (implemented)
+A serialization format for `Fr` elements,
+where an element is represented by a byte array `b[]` of size 32 with the least significant byte (LSB) coming first.
+
+NOTE: other implementation(s) using this format: ark-bls12-381-0.4.0, blst-0.3.7.
+
+#### `FormatFrMsb` (implemented)
+A serialization scheme for `Fr` elements,
+where an element is represented by a byte array `b[]` of size 32 with the most significant byte (MSB) coming first.
+
+NOTE: other implementation(s) using this format: ark-bls12-381-0.4.0, blst-0.3.7.
+
+#### `HashG1XmdSha256SswuRo` (implemented)
+The hash-to-curve suite `BLS12381G1_XMD:SHA-256_SSWU_RO_` that hashes a byte array into `G1` elements.
+Full specification is defined in https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#name-bls12-381-g1.
+
+#### `HashG2XmdSha256SswuRo` (implemented)
+The hash-to-curve suite `BLS12381G2_XMD:SHA-256_SSWU_RO_` that hashes a byte array into `G2` elements.
+Full specification is defined in https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16#name-bls12-381-g2.
+
+## Reference Implementation
+
+[https://github.com/aptos-labs/aptos-core/pull/6550/files](https://github.com/aptos-labs/aptos-core/pull/6550/files)
+
+## Risks and Drawbacks
+
+For move application developers, constructing cryptographic schemes manually with these building blocks can be error-prone, or even result in vulnerable applications.
+
+## Future Potential
+
+The module can be extended to support more structures and operations, allowing more complicated cryptographic applications to be built.
```

Comment:
Editorial: The `crypto_algebra.move` Move module can be extended to support more structures (e.g., new elliptic curves) and operations (e.g., batch inversion of field elements). This will:
1. Allow porting existing generic cryptosystems built on top of this module to new, potentially-faster-or-smaller curves.
2. Allow building more complicated cryptographic applications that leverage new operations or new algebraic structures (e.g., rings, hidden-order groups, etc.)

Reply:
Fixed.